### PR TITLE
Aip-on-azure

### PR DESCRIPTION
**Pull request by violetbrina**

### CODE REVIEW

File: cpg_infra/abstraction/azure.py [hunk 1]
```diff
@@ -2,48 +2,187 @@
...
+AZURE_BILLING_EXPIRY_DATE = "3141-25-09T00:00:00Z"
 class AzureInfra(CloudInfraBase):
     def __init__(
         self, config: CPGInfrastructureConfig, dataset_config: CPGDatasetConfig
     ):
         super().__init__(config, dataset_config)
 
-        self.resource_group_name = f'{config.dataset_storage_prefix}{self.dataset}'
-        self.storage_account_name = f'{config.dataset_storage_prefix}{self.dataset}'
+        self.region = 'australiaeast'
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:29]: Can we make this configurable?

File: cpg_infra/abstraction/azure.py [hunk 2]
```diff
@@ -2,48 +2,187 @@
...
+                    start_date=AZURE_BILLING_START_DATE,
+                    end_date=AZURE_BILLING_EXPIRY_DATE
+                ),
+                filter=filters
+            )
         )
 
     def bucket_rule_undelete(self, days=UNDELETE_PERIOD_IN_DAYS) -> Any:
-        pass
+        az.storage.BlobServiceProperties(
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:148]: Return this?

File: cpg_infra/abstraction/azure.py (hunk 1 above)
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:30]: We certainly can, but the gcp implementation currently doesn't either. Should we change both?

File: cpg_infra/abstraction/azure.py (hunk 2 above)
Comment (illusional) [reply cpg_infra/abstraction/azure.py:179]: This needs a comment on why it's not being returned - or the method needs to be renamed? What happens if you call it twice with different days of the week - is it referenced specifically later?

File: cpg_infra/abstraction/azure.py [hunk 3]
```diff
@@ -22,28 +29,200 @@ def __init__(
...
+            blob_services_name='default',
+            delete_retention_policy=az.storage.DeleteRetentionPolicyArgs(
+                days=days, enabled=True
+            ),
+            resource_group_name=self.resource_group.name,
         )
 
     def bucket_rule_undelete(self, days=UNDELETE_PERIOD_IN_DAYS) -> Any:
-        pass
+        self.storage_account_undelete_rule = True
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:190]: Need a follow-up comment here.

File: cpg_infra/abstraction/azure.py [hunk 4]
```diff
@@ -54,43 +233,73 @@ def create_bucket(
...
         return az.storage.BlobContainer(
-            f'bucket-{name}',
-            resource_group_name=self.resource_group_name,
-            account_name=self.storage_account_name,
+            f'{name}',
+            account_name=self.storage_account.name,
+            resource_group_name=project or self.resource_group.name,
             container_name=name,
+            metadata={'bucket': name},
+            public_access=az.storage.PublicAccess.BLOB,
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:263]: Can we turn off public access?

File: cpg_infra/abstraction/azure.py [hunk 5]
```diff
@@ -54,43 +233,73 @@ def create_bucket(
...
         self, resource_key: str, bucket, member, membership
     ) -> Any:
-        az.authorization.RoleAssignment(
-            resource_key,
-            scope=bucket.id,
-            principal_id=member.id,
-            role_definition_id='Contributor',
-            role_assignment_name='Storage Blob Data Contributor',
-        )
+        pass
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:270]: Any luck with this?

**Review by illusional — COMMENTED** [review]
Looking really great! One (very annoying) thing that I hadn't realised, is that if we have pulumi resource names that are the same across different infrastructures, Pulumi will probably be confused about which resource it should create.

It's very annoying, but could you add an `azure-` prefix to all of the resources that this creates?

File: stack/__main__.py [hunk 6]
```diff
@@ -18,12 +18,18 @@ def from_pulumi():
...
     config = CPGInfrastructureConfig.from_dict(get_config())
-
     pconfig = pulumi.Config()
     dataset_config = CPGDatasetConfig.from_pulumi(pconfig, dataset=pulumi.get_stack())
-
     CpgDatasetInfrastructure.deploy_all_from_config(config, dataset_config)
 
 
 if __name__ == '__main__':
+    # import debugpy
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply stack/__main__.py:27]: Sure thing

File: stack/cpg.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [infrastructure]
 domain = "populationgenomics.org.au"
-dataset_storage_prefix = "cpg-"
+dataset_storage_prefix = "cpg"
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply stack/cpg.toml:3]: Done!

File: cpg_infra/abstraction/azure.py [hunk 8]
```diff
@@ -22,28 +29,200 @@ def __init__(
     ):
         super().__init__(config, dataset_config)
 
-        self.resource_group_name = f'{config.dataset_storage_prefix}{self.dataset}'
-        self.storage_account_name = f'{config.dataset_storage_prefix}{self.dataset}'
+        self.region = 'australiaeast'
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:32]: In the cpg.toml or for each data set?

File: cpg_infra/abstraction/azure.py [hunk 9]
```diff
@@ -22,28 +29,200 @@ def __init__(
...
+            resource_group_name=self.resource_group.name,
+            policy=az.storage.ManagementPolicySchemaArgs(
+                rules=self.storage_account_lifecycle_rules,
+            ),
+            management_policy_name='default',
+        )
+
+    def create_project(self, name):
+        # TODO: re-work creating shared projects in Azure
+        return az.resources.ResourceGroup(name)
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:84]: This function gets called, so raising an error will cause the stack to fail. Creating the resource group works fine, we're just not sure if that is the final implemenation that we want for shared groups. We'll re-word the TODO instead.

File: cpg_infra/abstraction/azure.py (hunk 3 above)
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:190]: Done

File: cpg_infra/abstraction/azure.py (hunk 4 above)
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:263]: Done

File: cpg_infra/abstraction/azure.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py:34]: It's also hard coded in gcp.py. Should we change that too?

File: cpg_infra/abstraction/azure.py (hunk 5 above)
Comment (violetbrina) [reply cpg_infra/abstraction/azure.py]: Yup, working now!

File: cpg_infra/abstraction/azure.py [hunk 10]
```diff
@@ -54,43 +246,88 @@ def create_bucket(
...
+            account_name=self.storage_account.name,
+            resource_group_name=project or self.resource_group.name,
             container_name=name,
+            metadata={'bucket': name},
+            # TODO: work out requester_pays in Azure
         )
 
+    def bucket_membership_to_role(self, membership: BucketMembership):
+        if membership in (BucketMembership.MUTATE, BucketMembership.APPEND):
+            return f'{self.subscription}/providers/Microsoft.Authorization/roleDefinitions/ba92f5b4-2d11-453d-a403-e96b0029c9fe'
```
Comment (illusional) [inline cpg_infra/abstraction/azure.py:281]: Where does the last guid come from, could we make it a constant or add a comment?

**Review by illusional — COMMENTED** [review]
Looking really great! I've left a few comments, and might be worth a merge with main.
When you're very close to have it ready, we should try to deploy fewgenomes across GCP and Azure, and see what issues / clashes we see :)

**Review by illusional — APPROVED** [review]
Awesome 🎉

File: stack/new_stack.py [hunk 11]
```diff
@@ -574,32 +688,15 @@ def create_stack(
...
-        )
-    else:
-        logging.info(
-            f"""
+    logging.info(
+        f"""
 Created stack {dataset}, you can commit and push this with:
 
     git checkout -b add-{dataset}-stack
     git add {' '.join(files_to_add)}
```
Comment (vladsavelyev) [inline stack/new_stack.py:696]: Missing `git commit` here?

**Pull request merged**